src/worker/browser-env/runner/test-runner/index.ts

```diff
@@ -59,15 +67,114 @@ export class TestRunner extends NodejsEnvTestRunner {
 
     async run(opts: WorkerTestRunnerRunOpts): Promise<WorkerRunTestResult> {
         this._runOpts = opts;
+        let error: Error | undefined = undefined;
 
-        const results = await super.run({ ...opts, ExecutionThreadCls: wrapExecutionThread(this._socket) });
+        try {
+            await super.prepareToRun(this._runOpts);
 
+            error = await this._runWithAbort<Error | undefined>(throwIfAborted => {
+                const ExecutionThreadCls = wrapExecutionThread(this._socket, throwIfAborted);
+
+                return super.runRunnables(ExecutionThreadCls) as Promise<Error | undefined>;
+            });
+        } catch (err) {
+            error = err as Error;
+        }
+
+        while (error instanceof AbortError) {
+            error = undefined;
+
+            try {
+                await this._waitBroInitOnReconnect();
+
+                error = await this._runWithAbort<Error | undefined>(throwIfAborted => {
+                    const ExecutionThreadCls = wrapExecutionThread(this._socket, throwIfAborted);
+
+                    return new Promise((resolve, reject) => {
+                        (super.runRunnables(ExecutionThreadCls) as Promise<Error | undefined>)
+                            .then(resolve)
+                            .catch(reject);
+                    });
+                });
+            } catch (err) {
+                error = err as Error;
+            }
+        }
+
+        const results = await super.finishRun(error);
         this._socket.emit(WorkerEventNames.finalize);
 
+        if (error) {
+            throw error;
+        }
+
         return results;
     }
 
+    private _runWithAbort<T>(cb: (throwIfAborted: () => void) => Promise<T>): Promise<T> {
+        const controller = new AbortController();
+        const { signal } = controller;
+
+        let isAborted = signal.aborted;
+
+        signal.addEventListener("abort", () => {
+            isAborted = true;
+        });
+
+        const throwIfAborted = (): void => {
+            if (isAborted) {
+                throw new AbortError();
+            }
+        };
```

Comment:
Wouldn't `isAborted` be closured by value?

I know we could closure `signal`, since it is passed by reference:
`signal.aborted = true`, `if (signal.aborted)`
But im not sure about the value part